lib/core.js

```diff
@@ -196,6 +210,96 @@ function Core(options)
       JSON: {value: nativeTypeValues.JSON}
     });
 
+  /**
+  * Factory for Snowflake connections based on Generic Pool
+  *
+  * @param {Object} connectionOptions
+  *
+  * @returns {null}
+  */
+  function ConnectionFactory(connectionOptions)
+  {
+    /**
+     * Creates a new connection instance.
+     *
+     * @returns {Object}
+     */
+    this.create = function ()
+    {
+      const connection = new createConnection(connectionOptions);
+
+      return new Promise((resolve, reject) =>
+      {
+        connection.connect(
+          function (err, conn)
+          {
+            if (err)
+            {
+              console.error('Unable to connect: ' + err.message);
+              reject(new Error(err.message));
+            }
+            else
+            {
+              resolve(conn);
+            }
+          }
+        );
+      });
+    }
+
+    /**
+    * Destroys the specified connection instance.
+    *
+    * @param {Object} connection
+    *
+    * @returns {Object}
+    */
+    this.destroy = function (connection)
+    {
+      return new Promise((resolve) =>
+      {
+        connection.destroy(function (err, conn)
+        {
+          if (err)
+          {
+            console.error('Unable to disconnect: ' + err.message);
+          }
+          resolve();
+        });
+      });
+    }
+
+    /**
+    * Returns the status of the connection.
+    *
+    * @param {Object} connection
+    *
+    * @returns {Boolean}
+    */
+    this.validate = function (connection)
+    {
+      return connection.isUp();
```

Comment:
I don't think this is sufficient. The connection could be completely expired (both master and session token are expired) and `connection.isUp()` would return true (eventhough it should be false) since we haven't tried to make a request with the expired keys so we haven't moved into a new connection state. In the validate function you'll have to ensure that not only has the connection connected before, but that the session for that connection is still valid

Reply:
_"In the validate function you'll have to ensure that not only has the connection connected before"_
 - In the create function, a new connection instance will attempt to connect so each connection in the pool will have connected before
 
_"but that the session for that connection is still valid"_
 - Is the new isTokenValid() enough? It checks if both session token and master token's time are greater than current time